packages/kit/src/utils/routing.js

```diff
@@ -143,3 +149,20 @@ export function exec(match, { names, types, optional }, matchers) {
 
 	return params;
 }
+
+/** @param {string} str */
+function escape(str) {
+	return (
+		str
+			.normalize()
+			// escape [ and ] before escaping other characters, since they are used in the replacements
+			.replace(/[[\]]/g, '\\$&')
+			// replace %, /, ? and # with their encoded versions
+			.replace(/%/g, '%25')
```

Comment:
I'm a bit suspicious about this line and suspect perhaps it should be deleted. Can we create a test that checks `%` as a route id?

Reply:
it's definitely necessary. added a test, and also added it to the list of documented escape codes which was an oversight previously

Reply:
thanks. I guess I'll trust the fact that the test is passing, but I must admit I'm confused by this line still. The three lines after this make sense to me because `decodeURI` will leave those escape sequences untouched. However, it will turn `%25` into `%` which we'll then match against. So if we're matching against just `%` then it shouldn't be `%25` in the regex. Where did I go wrong? I'm wondering if there's something special with `%` that deserves an extra comment here

Reply:
ohhhh. I just figured out what is going on here. https://github.com/sveltejs/kit/pull/7521 special cased `%25`: https://github.com/sveltejs/kit/blob/7a7abbe34713813d3771aac21adb5ae8d1e66336/packages/kit/src/utils/url.js#L62

that's a pretty weird implementation. I suspect it might break `%2525`. If we leave it like this then we should probably add a test for that case. But I would personally instead change that implementation in `url.js` to something like this:
```
/**
 * @param {string} pathname
 */
export function decode_pathname(pathname) {
	// replace / with ? because / will screw up regex
	// there can't be a ? in the path so use it as a safe character and swap it back in decode_params
	return pathname.split('/').map(segment => decodeURIComponent(segment).replace(/\//g, '?')).join('/');
}

/** @param {Record<string, string>} params */
export function decode_params(params) {
	for (const key in params) {
		params[key] = params[key].replace(/\?/g, '/');
	}
	return params;
}
```

Reply:
hmm. maybe the current implementation is fine actually. but i'll suggest an update to the comment
```